### ISSUE TITLE
Create coding style document

### ISSUE THREAD

**Issue by rsanchezsaez** [ResearchKit/ResearchKit#174]
I took the liberty to draft a [ResearchKit Coding Style](https://github.com/rsanchezsaez/ResearchKit/wiki/ResearchKit-Coding-Style) document.

It mainly deals with visual formatting (whitespace, newlines, alignment, etc.). I think it can be useful to maintain code homogenization.

What do you think?

**Comment by brucehappy** [comment]
Yes, please.  I end up looking in other files, and find inconsistencies, so having a master document is wonderful.

**Comment by jwe-apple** [comment]
Seems useful. Do get a PR together, and please send an email to researchkit-dev as well to invite comment.

I'll want to tone down things like requiring alphabetical order and exact numbers of lines of whitespace between classes etc. The threshold for this is whether you'd feel it was too nitpicky when commenting on a PR for a new feature. But otherwise this seems like a great contribution - thanks for the effort!

**Comment by rsanchezsaez** [comment]
I'm afraid GitHub doesn't support PR of wiki pages. Either you'll have to copy and paste it into this project's wiki, or I can submit it as a regular markdown file inside the repo.
---
I've removed the alphabetical order requirement, that was silly.

However, I think it's useful to enforce a consistent number of empty newlines among the fules, the rules are not that hard to follow I'd say. Otherwise you end with arbitrary spacing in different files.
---
I'll send a mail to the list just now.

**Comment by rsanchezsaez** [comment]
@jwe-apple: By the way, I noticed that I missed a few files in my previous whitespace standardization PR (sorry about that). They're only the recorder subclasses. Could I possibly do these as well? Thanks.

**Comment by p2** [comment]
Good to have indeed! I can live with the spaces-for-indentation, one question about the _opening bracket_:

Agreed that it should go on the same line, but maybe **except** for top level method declarations? I'm used to seeing those after a newline (in other 3rd party libraries as well as Apple code), improves readability quite a bit. So the example would be:

```
// DO
- (void)doSomethingWithString:(NSString *)string
{
    if (condition) {
        ...
    } else {
        ...
    }
}
```

**Comment by rsanchezsaez** [comment]
@p2: Thanks for your comment.

Regarding braces, I could go with either style.

The problem is that all of the *ResearchKit core* codebase was already in the *opening brace on the same line* format, even for method definitions. So, adopting this would require an extensive reformatting (simple to do in an automated way, but potentially introducing *git-blame* noise).

I reckon that AppCore is in the style that you mentioned. It's mildly unfortunate that *AppCore* and *ResearchKit* don't use the same exact conventions.

On my own projects I used to use *opening brace on their own line*, but since diving into ResearchKit, *opening brace on the same line* has really grown on me. :smile:

**Comment by jwe-apple** [comment]
I'd prefer to keep opening brace on same line as the default going forward.

**Comment by p2** [comment]
Can't say I like the opening-method-brace-on-same-line because it makes scanning for a method name harder, but I can certainly agree to use it. ;)
